File: 백준/Gold/1717. 집합의 표현/집합의 표현.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.StringTokenizer;

public class Main {
    static int[] parent;

    public static void union(int a, int b){
        a = findParent(a);
        b = findParent(b);
        if(a != b) {
            parent[b] = a;
        }
    }

    public static int findParent(int a){
        if(a == parent[a])
            return a;
        else
            return parent[a] = findParent(parent[a]);
    }

    public static boolean hasSameParent(int a, int b){
        return findParent(a) == findParent(b);
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        parent = new int[N + 1];
        for (int i = 1; i <= N; i++) {
            parent[i] = i;
        }

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int command = Integer.parseInt(st.nextToken());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            if (command == 0) {
                union(a, b);
            }else {
                sb.append(hasSameParent(a, b) ? "YES" : "NO").append("\n");
            }
        }

        bw.write(sb.toString());
        bw.flush();
        bw.close();
        br.close();
    }
}

File: 백준/Gold/1717. 집합의 표현/집합의 표현.py
n, m = map(int, input().split())

parent = [i for i in range(n+1)]


def find(x):
    if parent[x] != x:
        parent[x] = find(parent[x])
    return parent[x]


def union(x, y):
    root_x = find(x)
    root_y = find(y)

    if root_x != root_y:
        parent[root_y] = root_x


for _ in range(m):
    command, a, b = map(int, input().split())
    if command == 0:
        union(a, b)
    else:
        root_a = find(a)
        root_b = find(b)
        if root_a == root_b:
            print("YES")
        else:
            print("NO")
            